src/ScriptRuntime/Execution/ProcessExecutor.php

```diff
@@ -0,0 +1,379 @@
+<?php declare(strict_types=1);
+
+
+namespace Shopware\Psh\ScriptRuntime\Execution;
+
+use Shopware\Psh\Listing\Script;
+use Shopware\Psh\ScriptRuntime\BashCommand;
+use Shopware\Psh\ScriptRuntime\Command;
+use Shopware\Psh\ScriptRuntime\DeferredProcessCommand;
+use Shopware\Psh\ScriptRuntime\ParsableCommand;
+use Shopware\Psh\ScriptRuntime\ProcessCommand;
+use Shopware\Psh\ScriptRuntime\SynchronusProcessCommand;
+use Shopware\Psh\ScriptRuntime\TemplateCommand;
+use Shopware\Psh\ScriptRuntime\WaitCommand;
+use Symfony\Component\Process\Process;
+
+/**
+ * Execute a command in a separate process
+ */
+class ProcessExecutor
+{
+    /**
+     * @var ProcessEnvironment
+     */
+    private $environment;
+
+    /**
+     * @var TemplateEngine
+     */
+    private $templateEngine;
+
+    /**
+     * @var Logger
+     */
+    private $logger;
+
+    /**
+     * @var string
+     */
+    private $applicationDirectory;
+
+    /**
+     * @var DeferredProcess[]
+     */
+    private $deferredProcesses = [];
+
+    /**
+     * ProcessExecutor constructor.
+     * @param ProcessEnvironment $environment
+     * @param TemplateEngine $templateEngine
+     * @param Logger $logger
+     * @param string $applicationDirectory
+     */
+    public function __construct(
+        ProcessEnvironment $environment,
+        TemplateEngine $templateEngine,
+        Logger $logger,
+        string $applicationDirectory
+    ) {
+        $this->environment = $environment;
+        $this->templateEngine = $templateEngine;
+        $this->logger = $logger;
+        $this->applicationDirectory = $applicationDirectory;
+    }
+
+    /**
+     * @param Script $script
+     * @param Command[] $commands
+     */
+    public function execute(Script $script, array $commands)
+    {
+        $this->logger->startScript($script);
+
+        $this->executeTemplateRendering();
+
+        try {
+            foreach ($commands as $index => $command) {
+                $this->executeCommand($command, $index, count($commands));
+            }
+        } finally {
+            $this->waitForDeferredProcesses();
+        }
+
+        $this->logger->finishScript($script);
+    }
+
+    /**
+     * @param Command $command
+     * @param int $index
+     * @param int $totalCount
+     */
+    private function executeCommand(Command $command, int $index, int $totalCount)
+    {
+        switch (true) {
```

Comment:
Warum benutzt du `switch(true)` und dann in den cases jeweils `instanceof` anstelle den FQCN zu vergleichen?

`switch($command::class)` ... `case BashCommand::class`

Reply:
Intern wird es dann ein String-Vergleich und damit rennt man in die `==` vs `===` Problematik. Man wünscht sich also das zweite, bekommt aber das erste ;)